File: Data Structures & Algorithms/top-k-elements-in-list/submission-2.cpp
#include <vector>
#include <array>
#include <algorithm>

class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        std::array<int, 2001> freq = {0};
        for(const int n:nums)
            freq[n+1000]++;

        vector<pair<int, int>> counts;
        counts.reserve(2001);
        for(int i=0; i<2001; i++)
        {
            if(freq[i]>0)
                counts.push_back({freq[i], i-1000});
        }
        
        sort(counts.begin(), 
        counts.end(),
        [](const auto&a, const auto&b){return a.first>b.first;});

        vector<int> res;
        res.reserve(k);
        for(int i=0;i<k;i++)
            res.push_back(counts[i].second);
        
        return res;
    }
};

File: Data Structures & Algorithms/three-integer-sum/submission-2.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums, l = sorted(nums), len(nums)
        for i in range(l):
            if i>0 and nums[i]==nums[i-1]:
                continue
            t = -nums[i]
            j,k = i+1, l-1
            while j<k:
                s = nums[j]+nums[k]
                if s==t:
                    res.append([nums[i],nums[j],nums[k]])
                    j+=1
                    k-=1

                    while j<k and nums[j]==nums[j-1]:
                        j+=1

                elif s<t:
                    j+=1
                else:
                    k-=1
        return res



                